frontend/apps/reader/modules/readersearch.lua

```diff
@@ -51,33 +83,103 @@ function ReaderSearch:onShowFulltextSearchInput()
                     text = backward_text,
                     callback = function()
                         if self.input_dialog:getInputText() == "" then return end
-                        UIManager:close(self.input_dialog)
-                        self:onShowSearchDialog(self.input_dialog:getInputText(), 1)
+                        self.last_search_text = self.input_dialog:getInputText()
+                        self.use_regex = self.check_button_regex.checked
+                        self.case_insensitive = not self.check_button_case.checked
+                        if self.use_regex and self.ui.document:checkRegex(self.input_dialog:getInputText()) ~= 0 then
+                            UIManager:show(InfoMessage:new{ text = _("Error in regular expression!") })
+                        else
+                            UIManager:close(self.input_dialog)
+                            self:onShowSearchDialog(self.input_dialog:getInputText(), 1, self.use_regex, self.case_insensitive)
+                        end
                     end,
                 },
                 {
                     text = forward_text,
                     is_enter_default = true,
                     callback = function()
                         if self.input_dialog:getInputText() == "" then return end
-                        UIManager:close(self.input_dialog)
-                        self:onShowSearchDialog(self.input_dialog:getInputText(), 0)
+                        self.last_search_text = self.input_dialog:getInputText()
+                        self.use_regex = self.check_button_regex.checked
+                        self.case_insensitive = not self.check_button_case.checked
+                        if self.use_regex and self.ui.document:checkRegex(self.input_dialog:getInputText()) ~= 0 then
+                            UIManager:show(InfoMessage:new{ text = _("Error in regular expression!") })
+                        else
+                            UIManager:close(self.input_dialog)
+                            self:onShowSearchDialog(self.input_dialog:getInputText(), 0, self.use_regex, self.case_insensitive)
+                        end
                     end,
                 },
             },
         },
     }
+
+   -- checkboxes
+    self.check_button_regex = CheckButton:new{
+        text = _("Regular expression"),
+        face = Font:getFace("smallinfofont"),
+        checked = self.use_regex,
+        parent = self.input_dialog,
+        callback = function()
+            if not self.check_button_regex.checked then
+                self.check_button_regex:check()
+            else
+                self.check_button_regex:unCheck()
+            end
+        end,
+        padding = self.input_dialog.padding,
+        margin = self.input_dialog.margin,
+        bordersize = self.input_dialog.bordersize,
```

Comment:
Looks like these are not even used in CheckButton. Can you try removing them and see if things change? If no, just remove them to make this shorter.

Reply:
Seems to work without.